tensorflow/core/kernels/crop_and_resize_op.cc

```diff
@@ -442,70 +443,92 @@ struct CropAndResizeBackpropImage<CPUDevice, T> {
-              const float dbottom = y_lerp * grads(b, y, x, d);
-              grads_image(b_in, bottom_y_index, left_x_index, d) +=
-                  static_cast<T>((1 - x_lerp) * dbottom);
-              grads_image(b_in, bottom_y_index, right_x_index, d) +=
-                  static_cast<T>(x_lerp * dbottom);
+          const int top_y_index = floorf(in_y);
+          const int bottom_y_index = ceilf(in_y);
+          const float y_lerp = in_y - top_y_index;
+
+          for (int x = 0; x < crop_width; ++x) {
+            const float in_x = (crop_width > 1)
+                                   ? x1 * (image_width - 1) + x * width_scale
+                                   : 0.5 * (x1 + x2) * (image_width - 1);
+            if (in_x < 0 || in_x > image_width - 1) {
+              continue;
             }
-          } else {  // method_name == "nearest"
-            for (int d = 0; d < depth; ++d) {
-              int closest_x_index = roundf(in_x);
-              int closest_y_index = roundf(in_y);
-              grads_image(b_in, closest_y_index, closest_x_index, d) +=
-                  static_cast<T>(grads(b, y, x, d));
+
+            if (method_name == "bilinear") {
+              const int left_x_index = floorf(in_x);
+              const int right_x_index = ceilf(in_x);
+              const float x_lerp = in_x - left_x_index;
+
+              for (int d = 0; d < depth; ++d) {
+                const float dtop = (1 - y_lerp) * grads(b, y, x, d);
+                grads_image(b_in, top_y_index, left_x_index, d) +=
+                    static_cast<T>((1 - x_lerp) * dtop);
+                grads_image(b_in, top_y_index, right_x_index, d) +=
+                    static_cast<T>(x_lerp * dtop);
+                const float dbottom = y_lerp * grads(b, y, x, d);
+                grads_image(b_in, bottom_y_index, left_x_index, d) +=
+                    static_cast<T>((1 - x_lerp) * dbottom);
+                grads_image(b_in, bottom_y_index, right_x_index, d) +=
+                    static_cast<T>(x_lerp * dbottom);
+              }
+            } else {  // method_name == "nearest"
+              for (int d = 0; d < depth; ++d) {
+                int closest_x_index = roundf(in_x);
+                int closest_y_index = roundf(in_y);
+                grads_image(b_in, closest_y_index, closest_x_index, d) +=
+                    static_cast<T>(grads(b, y, x, d));
+              }
             }
           }
         }
       }
-    }
+    };
+
+    // A rough estimation of the cost for each cropped box.
+    // Including calculation cost in the depth loop and pixel loop.
+    const double cost_per_pixel =
+        (method_name == "bilinear"
+             ? depth * (Eigen::TensorOpCost::AddCost<float>() * 4 +
+                        Eigen::TensorOpCost::MulCost<float>() * 6 +
+                        Eigen::TensorOpCost::CastCost<T, float>() * 4) +
+                   Eigen::TensorOpCost::AddCost<float>() * 4
```

Comment:
I only found three additions, `in_x - left_x_index`, `1 - y_lerp` , and `1 - x_lerp`. Where is the 4th one?

Reply:
My best guess is:
1. line 467:
    <pre>
    const float in_y = (crop_height > 1)
                                 ? y1 * (image_height - 1) + y * height_scale
                                 : 0.5 * (y1 + y2) * (image_height - 1);
    </pre>
    both branch has 2 additions;
2. line 481:
    <pre>
    if (in_x < 0 || in_x > image_width - 1) {
    </pre>
    That's 1 addition;
3. line 488:
    <pre>
    const float x_lerp = in_x - left_x_index;
    </pre>
    That's 1 addition.

Reply:
And I suppose you count `1 - y_lerp` and `1 - x_lerp` as non-loop additions because you assume compiler will do loop invariant motion for them? But don't we not consider compiler optimizations in cost model, because cost model is merely a rough estimation? I don't know.

Reply:
So sorry for my late reply! I only looked for addition in the `bilinear` if case because you only added 4 `AddCost`s to the `bilinear` cost and not to `nearest`. In the `if` there are 4 `+=` (at the depth level) so I guessed you counted `1 - y_lerp` and `1 - x_lerp` as loop invariants. I don't usually try to count for compiler optimizations either. :) In the end this is just a rough estimate to help shard the work better. I agree about ignoring optimizations.

I didn't author this code too. How about listing the costs explicitly here and see if there is anything missing? I'll start:

Cost at the pixel level (both methods):
2 additions and 2 multiplications for either branch.
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L478-L480
1 addition from
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L481

Cost at the pixel level for just `bilinear`:
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L488

Cost at the depth level for `bilinear`:
3 additions for `1 - y_lerp` and `1 - x_lerp`
4 additions for `+=`
6 multiplications

Cost at the depth level for `nearest`
1 addition and 1 cast.

Anything else?

Reply:
I only looked for addition in the `bilinear` if case because you only added 4 `AddCost`s to the `bilinear` cost and not to `nearest`. In the `if` there are 4 `+=` (at the depth level) so I guessed you counted `1 - y_lerp` and `1 - x_lerp` as loop invariants. I don't usually try to count for compiler optimizations either. :) In the end this is just a rough estimate to help shard the work better. I agree about ignoring optimizations.

I didn't author this code too. How about listing the costs explicitly here and see if there is anything missing? I'll start:

Cost at the pixel level (both methods):
2 additions and 2 multiplications for either branch.
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L478-L480
1 addition from
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L481

Cost at the pixel level for just `bilinear`:
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L488

Cost at the depth level for `bilinear`:
3 additions for `1 - y_lerp` and `1 - x_lerp`
4 additions for `+=`
6 multiplications

Cost at the depth level for `nearest`
1 addition and 1 cast.

Anything else?

Reply:
Generally I agree. However I think there are 4 casts in `bilinear` depth level.

So that's our conclusion:
1. for `bilinear`:
    4 addition + depth * (7 addition + 6 multiplication + 4 cast)
2. for 'nearest':
    3 addition + depth * (1 addition + 1 cast)

Reply:
Right. I forgot about the casts in bilinear. Thank you!